MTV-2678 | RFE: Implement OpenShift Network Policies in MTV

Hazanel opened the pull request:
Ref: https://issues.redhat.com/browse/MTV-2678
 Signed-off-by: Elad Hazan <ehazan@redhat.com>



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Adds a configurable NetworkPolicies feature (off by default) to manage ingress/egress for operator, controller, API, migration workloads, and validation components.
  - Provides templated NetworkPolicy manifests for operator, controller, API, migration workloads, and validation components.
  - Adds optional NFSv3 port support behind a toggle and automatic cleanup of NetworkPolicy resources when the feature is disabled.

- **Chores**
  - Grants RBAC permissions to manage networking.k8s.io/networkpolicies.
  - Surfaces the feature flag across sample and installation manifests (including CSVs) for easier configuration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Hazanel committed:
Resolves: MTV-2678 | RFE: Add network policies fixes and version-aware enablement logic
Ref: https://issues.redhat.com/browse/MTV-2678
Signed-off-by: Elad Hazan <ehazan@redhat.com>

Hazanel commented:
<img width="3840" height="860" alt="Network_policies_diagram" src="https://github.com/user-attachments/assets/cb1f3ebc-ab8e-452f-8925-7577064e0861" />

Hazanel commented on operator/config/network-policies/README.md:
1. This README isn't necessarily for users but we can discuss its location.
2. Those yaml files are the actual network-policies that should be enabled with the new flag in the spec

yaacov replied:
> we can discuss its location

the doc file and the yamls in the doc dir, are in the code dir,

> yaml files are the actual network-policies 

you have the jinja templates in a different dir for deployment, AFAIU the yaml files in the `operator/config/network-policies` are part of the documentation, what am i missing ?

mnecas reviewed (commented):
Where did you got the list of those ports?
Have you tried the migrations from all providers?
This can break all of the migraitons if we miss one port.

Hazanel commented:
@mnecas 
1.  Tested (with the standalone files who emulates the j2 templates file)a full creation and migration plan with vsphere, ova and ocp
2. 
Port    Protocol    Component/Usage                               Purpose
53      UDP/TCP     All Pods                                                  DNS Resolution
80      TCP              Migration/Validation Workloads       HTTP (OVA downloads, web sources)
443     TCP            All Components                                    HTTPS (Kubernetes API, Providers, OVA downloads)
902     TCP            Migration/Validation Workloads        vSphere NFC (Network File Copy)
2049    TCP/UDP     Migration/Validation Workloads     NFS (Network File System) Storage
5000    TCP           Migration Workloads                           OpenStack Keystone (Identity Service)
6443    TCP          All Components                                    Kubernetes API Server
8080    TCP         Migration Workloads                           oVirt Engine HTTP
8383    TCP         Forklift Operator                                   Prometheus Metrics
8443    TCP         All Components                                    Forklift API Service / Provider HTTPS Alternate
8444    TCP         Forklift API                                             Forklift API Metrics Endpoint
8774    TCP         Migration Workloads                           OpenStack Nova (Compute Service)
8776    TCP         Migration Workloads                           OpenStack Cinder (Block Storage)
9696    TCP         Migration Workloads                           OpenStack Neutron (Networking)
35357   TCP         Migration Workloads                          OpenStack Admin API
54322   TCP         Migration Workloads                          oVirt Imageio (Disk Transfer Service)

Hazanel commented on operator/roles/forkliftcontroller/defaults/main.yml:
```diff
@@ -19,6 +20,7 @@ forklift_resources:
   - ConfigMap
   - Service
   - Route
+  - NetworkPolicy
```
It's already gated - false alarm

mnecas commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-operator.yml.j2:
```diff
@@ -0,0 +1,48 @@
+        matchLabels:
+          kubernetes.io/metadata.name: openshift-monitoring
+    ports:
+    - protocol: TCP
+      port: 8383  # Metrics port
```
Where is this coming from?

mnecas replied:
@Hazanel don't resolve the comments without answering.

mnecas commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-validation.yml.j2:
```diff
@@ -0,0 +1,161 @@
+      port: 443   # vSphere API
+    - protocol: TCP
+      port: 8443  # vSphere HTTPS alternate
+    - protocol: TCP
+      port: 902   # NFC
```
Do we need this?

mnecas commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-migration-workloads.yml.j2:
```diff
@@ -0,0 +1,280 @@
+    - protocol: TCP
+      port: 8443    # oVirt HTTPS alternate
+    - protocol: TCP
+      port: 54322   # oVirt imageio
```
What about the port 54323 which of these ports is used?

mnecas commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-migration-workloads.yml.j2:
```diff
@@ -0,0 +1,280 @@
+    - protocol: TCP
+      port: 443     # oVirt API
+    - protocol: TCP
+      port: 8443    # oVirt HTTPS alternate
```
What alternative? The 8443 is used for development

mnecas commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-controller.yml.j2:
```diff
@@ -0,0 +1,74 @@
+      port: 8776  # OpenStack Cinder
+    - protocol: TCP
+      port: 9696  # OpenStack Neutron
+                - protocol: TCP
+              port: 35357 # OpenStack Admin (legacy; modern Keystone commonly uses 5000)
```
Formatting? This could not work. Have you tested it?

mnecas replied:
@Hazanel don't resolve the comments without answering.

mnecas commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-validation.yml.j2:
```diff
@@ -0,0 +1,161 @@
+      port: 2049  # NFS (NFSv4/NFSv3 data)
+    - protocol: UDP
+      port: 2049  # NFS (NFSv4/NFSv3 data)
+{% if nfs_v3_support|default(false)|bool %}
```
Where is this coming from?

mnecas commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-migration-workloads.yml.j2:
```diff
@@ -0,0 +1,280 @@
+    - protocol: TCP
+      port: 35357 # OpenStack Admin (legacy; modern Keystone commonly uses 5000)
+    - protocol: TCP
+      port: 8080  # oVirt Engine HTTP
```
8080? not 80 from HTTP?

mnecas commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-migration-workloads.yml.j2:
```diff
@@ -0,0 +1,280 @@
+spec:
+  podSelector:
+    matchLabels:
+      forklift.app: virt-v2v
```
This is used for v2v? does openstack use v2v or RHV?

mnecas commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-migration-workloads.yml.j2:
```diff
@@ -0,0 +1,280 @@
+    - protocol: TCP
+      port: 443   # HTTPS
+    - protocol: TCP
+      port: 8443  # HTTPS alternate / oVirt Engine HTTPS
```
8443 Dev not alternative

mnecas replied:
Nope, that's not my point.
We should not be using development ports of other products.

mnecas commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-validation.yml.j2:
```diff
@@ -0,0 +1,161 @@
+    # All ports allowed (ports omitted)
+  # Allow vSphere API access for VDDK validation
+  - ports:
+    - protocol: TCP
+      port: 443   # vSphere API
+    - protocol: TCP
+      port: 8443  # vSphere HTTPS alternate
+    - protocol: TCP
```
What is the vddk validation doing that it needs to have these ports?

mnecas reviewed (changes requested):
This is not easy problem, that's why it's epic.
Lot of these configurations I don't know from top of my head. You need to go through the Forklift code to understand what's going on and which policies we need and to which pod.  Also you will need to know the providers how we interact with them and which ports we need to have for them.
Large language models should not own the patch, you are responsible for the code.

Please don't close the comments without answering them.

Hazanel commented:
<img width="1844" height="616" alt="image" src="https://github.com/user-attachments/assets/36ef9535-55ac-45c5-8585-9169cda06fa9" />

Hazanel commented:
Validate .kustomized_manifests / validate  fails on windows legacy latest changes  - not related to this patch

Hazanel commented:
Port	Service / Purpose	Code Evidence	Network Policy
53	DNS (UDP/TCP)	Standard DNS resolution	All policies
80	HTTP Standard	❓ Not explicitly listed (covered by 8080)	N/A
443	HTTPS Standard	Standard HTTPS for provider APIs	All migration workload policies
902	vSphere NFC	vendor/github.com/vmware/govmomi/vim25/types/types.go:36329	forklift-virt-v2v-policy, forklift-vsphere-xcopy-populator-policy
2112	Operator Metrics	pkg/settings/metrics.go	forklift-operator-policy
5000	OpenStack Keystone	Standard OpenStack Keystone API	forklift-virt-v2v-policy, forklift-openstack-populator-policy
6443	Kubernetes API	Standard Kubernetes API port	forklift-controller-policy, forklift-api-policy
8080	OVA Provider + oVirt Metrics	cmd/ova-provider-server/ova-provider-server.go:211 + cmd/populator-controller/populator-controller.go:40	forklift-controller-policy, forklift-ovirt-populator-policy
8081	Controller + OpenStack Metrics	deployment-controller.yml.j2:49 + cmd/populator-controller/populator-controller.go:47	forklift-controller-policy, forklift-openstack-populator-policy
8082	vSphere Xcopy Metrics	cmd/populator-controller/populator-controller.go:54	forklift-vsphere-xcopy-populator-policy
8181	Validation Service (OPA)	pkg/controller/validation/policy/client.go:118-173	forklift-validation-service-policy, forklift-controller-policy
8443	API Service + HTTPS Dev	pkg/forklift-api/api.go + Common dev port	forklift-api-policy, All migration policies
8444	API Service Endpoint	pkg/forklift-api/api.go	forklift-api-policy
8774	OpenStack Nova	Standard OpenStack Nova API	forklift-virt-v2v-policy, forklift-openstack-populator-policy
8776	OpenStack Cinder	Standard OpenStack Cinder API	forklift-virt-v2v-policy, forklift-openstack-populator-policy
9696	OpenStack Neutron	Standard OpenStack Neutron API	forklift-virt-v2v-policy, forklift-openstack-populator-policy
35357	OpenStack Admin (Legacy)	Legacy OpenStack Keystone admin	forklift-virt-v2v-policy, forklift-openstack-populator-policy
54322	oVirt imageio-daemon	Standard oVirt imageio direct transfer	forklift-virt-v2v-policy, forklift-ovirt-populator-policy
54323	oVirt imageio-proxy	Standard oVirt imageio proxy transfer	forklift-virt-v2v-policy, forklift-ovirt-populator-policy

mnecas reviewed (changes requested):
Nice table, but you can't group the service ports together. Each service connects to differnt thing and ports are used for different reason, some of them are generic that could apply to most of them but not all.

How did you test it?
What exact steps did you do to test it?
How did you ensure that the policies are accually working?
Did you limit the policies and tried the communication?

It is hard to see changes what were done as this is huge PR.
Don't force push the same commit but add another commits, so I can review the changes.

What about the MTV hooks?

mnecas reviewed (changes requested):
I will NOT merge nor review 2000 line AI generated md file describing stuff how it was tested knowing that the basic tests could not work on our enviroments. As you are still missing critical ports. For example OpenStack 13000.

This PR is dagnerous as even one missing port can break the enviroment. This is not complex you need to go through the docs, test the changes. Even tho AI generated the code YOU are resposible for it.

Please follow RedHat AI Guidlines. https://source.redhat.com/projects_and_programs/ai/wiki/code_assistants_guidelines_for_responsible_use_of_ai_code_assistants
```
Assisted-by: <name of code assistant>
```
and
```
Generated by <name of code assistant>
```

Moving to draft.

mnecas commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-api.yml.j2:
```diff
@@ -0,0 +1,64 @@
+      - namespaceSelector: {}
+      ports:
+      - protocol: TCP
+        port: 8444
```
The UI sends the requests directly to the API not to through the service? This sounds strange

mnecas commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-validation.yml.j2:
```diff
@@ -0,0 +1,67 @@
+              service: {{ api_service_name }}
+      ports:
+        - protocol: TCP
+          port: 8181
```
hmm didn't know that we communicate from forklift-api to validation. Where is this in the code?

mnecas commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-migration-workloads.yml.j2:
```diff
@@ -0,0 +1,315 @@
+    - protocol: TCP
+      port: 902   # vSphere NFC storage transfer
+    - protocol: TCP
+      port: 8080  # OVA downloads HTTP
```
So this is for virt-v2v pod, how is it related to OVA download?

Hazanel replied:
That might be a better comment:
" Optional: allow TCP 8080 only if OVA/image sources are served over HTTP on nonstandard ports"

We're using it in the ova policy so I assumed we need it here as well,if you think it's not required I can drop that

Hazanel commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-migration-workloads.yml.j2:
```diff
@@ -0,0 +1,315 @@
+  - from:
+    - podSelector:
+        matchLabels:
+          forklift.app: migration-workload
```
right - it's virt-v2v

Hazanel commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-migration-workloads.yml.j2:
```diff
@@ -0,0 +1,315 @@
+    - protocol: TCP
+      port: 13774 # Custom OpenStack service port
+    - protocol: TCP
+      port: 13776 # Custom OpenStack service port
```
I was a bit hesitant about that as you remember - I'll open egress

Hazanel commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-operator.yml.j2:
```diff
@@ -0,0 +1,48 @@
+          kubernetes.io/metadata.name: openshift-monitoring
+    ports:
+    - protocol: TCP
+      port: 2112  # Conversion pod metrics
```
I assumed metrics should be exposed at the operator level as well

Hazanel commented on operator/roles/forkliftcontroller/templates/security/networkpolicy-validation.yml.j2:
```diff
@@ -0,0 +1,67 @@
+      forklift.app: vddk-validation
+  policyTypes:
+  - Ingress
+  - Egress
```
I dropped the policy

Hazanel committed:
Add some network policies fixes + version-aware network policies logic for enablement

Resolves: None
Signed-off-by: Elad Hazan <ehazan@redhat.com>

mnecas commented on operator/roles/forkliftcontroller/tasks/main.yml:
```diff
@@ -21,6 +21,56 @@
+        failed_when: false
+    when: not (feature_network_policies | default('false') | bool)
+    # Only set version-aware default if user hasn't explicitly enabled it in ForkliftController CR
+    # If user sets feature_network_policies: "true" in CR, that takes precedence
```
We already have `ocp_version` in the https://github.com/kubev2v/forklift/blob/993c1a80eaaacc439d38ef4edce01d2b40b58907/operator/roles/forkliftcontroller/tasks/main.yml#L49

Hazanel committed:
MTV-2678 | Network policies

- version-aware enablement refactor
- Added HyperV network policy

Resolves: None
Signed-off-by: Elad Hazan <ehazan@redhat.com>
